Frameworks/MQTTnet.NetCoreApp/MqttClientSslChannel.cs

```diff
@@ -0,0 +1,122 @@
+﻿using System;
+using System.IO;
+using System.Net.Security;
+using System.Net.Sockets;
+using System.Security.Authentication;
+using System.Threading.Tasks;
+using MQTTnet.Core.Channel;
+using MQTTnet.Core.Client;
+using MQTTnet.Core.Exceptions;
+
+namespace MQTTnet
+{
+    /// <summary>
+    /// Describes an SSL channel to an MQTT server.
+    /// </summary>
+    public class MqttClientSslChannel : IMqttCommunicationChannel, IDisposable
+    {
+        private readonly Socket _socket;
+        private SslStream _sslStream;
+
+        /// <summary>
+        /// Creates a new <see cref="MqttClientSslChannel"/>.
+        /// </summary>
+        public MqttClientSslChannel()
+        {
+            _socket = new Socket(SocketType.Stream, ProtocolType.Tcp);
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="MqttClientSslChannel"/> with a predefined <paramref name="socket"/>.
+        /// </summary>
+        /// <param name="socket"></param>
+        public MqttClientSslChannel(Socket socket)
+        {
+            _socket = socket ?? throw new ArgumentNullException(nameof(socket));
+        }
+
+        /// <summary>
+        /// Asynchronously connects to the host described in the <see cref="MqttClientOptions"/>.
+        /// </summary>
+        /// <param name="options">The <see cref="MqttClientOptions"/> describing the connection.</param>
+        public async Task ConnectAsync(MqttClientOptions options)
+        {
+            try
+            {
+                await _socket.ConnectAsync(options.Server, options.Port);
+
+                NetworkStream ns = new NetworkStream(_socket, true);
+                _sslStream = new SslStream(ns);
+
+                await _sslStream.AuthenticateAsClientAsync(options.Server, null, SslProtocols.Tls12, false);
```

Comment:
You can return the Task directly if no code follows the await. Because otherwise the context will be created and restored for nothing. It will increase the performance.